api/capi/src/nntrainer.cpp

```diff
@@ -492,12 +501,35 @@ int ml_train_model_set_dataset(ml_train_model_h model,
   }
 
   std::shared_ptr<ml::train::Model> m;
-  std::shared_ptr<ml::train::Dataset> d;
 
   m = nnmodel->model;
-  d = nndataset->dataset;
 
-  returnable f = [&]() { return m->setDataset(d); };
+  returnable f = [&]() {
+    auto &[train_set, valid_set, test_set] = nndataset->dataset;
+    int status = ML_ERROR_NONE;
+    status =
+      m->setDataset(ml::train::DatasetDataUsageType::DATA_TRAIN, train_set);
+    if (status != ML_ERROR_NONE) {
+      return status;
+    }
+
+    if (valid_set != nullptr) {
+      status =
+        m->setDataset(ml::train::DatasetDataUsageType::DATA_VAL, valid_set);
+      if (status != ML_ERROR_NONE) {
+        return status;
+      }
+    }
+
+    if (test_set != nullptr) {
+      status =
+        m->setDataset(ml::train::DatasetDataUsageType::DATA_TEST, test_set);
+      if (status != ML_ERROR_NONE) {
+        return status;
+      }
+    }
```

Comment:
Creating a dataset with (deprecated) capi will be split into three underlying calls.